cpp/include/cudf/strings/detail/gather.cuh

```diff
@@ -294,59 +293,31 @@ std::unique_ptr<cudf::column> gather(strings_column_view const& strings,
-  auto const d_in_offsets  = (strings_count > 0) ? strings.offsets_begin() : nullptr;
-  auto const d_strings     = column_device_view::create(strings.parent(), stream);
-  thrust::transform(
-    rmm::exec_policy_nosync(stream),
-    begin,
-    end,
-    d_out_offsets,
-    [d_strings = *d_strings, d_in_offsets, strings_count] __device__(size_type in_idx) {
-      if (NullifyOutOfBounds && (in_idx < 0 || in_idx >= strings_count)) return 0;
-      if (not d_strings.is_valid(in_idx)) return 0;
-      return d_in_offsets[in_idx + 1] - d_in_offsets[in_idx];
-    });
+  // build offsets column
+  auto const d_strings    = column_device_view::create(strings.parent(), stream);
+  auto const d_in_offsets = !strings.is_empty() ? strings.offsets_begin() : nullptr;
 
-  // check total size is not too large
-  size_t const total_bytes = thrust::transform_reduce(
-    rmm::exec_policy_nosync(stream),
-    d_out_offsets,
-    d_out_offsets + output_count,
-    [] __device__(auto size) { return static_cast<size_t>(size); },
-    size_t{0},
-    thrust::plus{});
-  CUDF_EXPECTS(total_bytes < static_cast<std::size_t>(std::numeric_limits<size_type>::max()),
-               "total size of output strings exceeds the column limit",
-               std::overflow_error);
-
-  // In-place convert output sizes into offsets
-  thrust::exclusive_scan(rmm::exec_policy_nosync(stream),
-                         d_out_offsets,
-                         d_out_offsets + output_count + 1,
-                         d_out_offsets);
+  auto offsets_itr = thrust::make_transform_iterator(
+    begin, [d_strings = *d_strings, d_in_offsets] __device__(size_type idx) {
+      if (NullifyOutOfBounds && (idx < 0 || idx >= d_strings.size())) { return 0; }
+      if (not d_strings.is_valid(idx)) { return 0; }
+      return d_in_offsets[idx + 1] - d_in_offsets[idx];
+    });
+  auto [out_offsets_column, total_bytes] =
+    cudf::detail::make_offsets_child_column(offsets_itr, offsets_itr + output_count, stream, mr);
 
   // build chars column
-  cudf::device_span<int32_t const> const d_out_offsets_span(d_out_offsets, output_count + 1);
-  auto out_chars_column = gather_chars(d_strings->begin<string_view>(),
-                                       begin,
-                                       end,
-                                       d_out_offsets_span,
-                                       static_cast<size_type>(total_bytes),
-                                       stream,
-                                       mr);
+  auto const offsets_view = out_offsets_column->view();
+  auto out_chars_column   = gather_chars(
+    d_strings->begin<string_view>(), begin, end, offsets_view, total_bytes, stream, mr);
 
   return make_strings_column(output_count,
                              std::move(out_offsets_column),
                              std::move(out_chars_column),
-                             0,
+                             0,  // caller sets these
```

Comment:
Design question: the caller is currently expected to set the validity mask of the output. Can you explain precisely what this means? Are null inputs currently replaced by empty strings in the output? Is the caller supposed to apply the same gather map to the input null mask, then assign that mask/count to the output column? It makes me wonder if we should use a fancier iterator to handle mask gathering at the same time as string gathering, reducing the number of kernel launches. Does that negatively impact performance?

Reply:
This is just a specialization of gather for this type.  There is some common code before/after the dispatcher that handles nulls. This specialization is not called without that dispatch logic. This particular implementation could build the nulls but there is no reason to do it twice. We could make each specialization handle the nulls but that could duplicate the bitmask computation across many functions.
I think it would be difficult to make a fancy iterator for gathering individual bits perform well over the current bitmask gather kernel.  Either way, I think this is a bigger discussion and out of scope for this PR.

Reply:
Great! That’s exactly what I needed to know. Thanks for the context.